amethyst_network/src/simulation/transport/tcp.rs

```diff
@@ -0,0 +1,276 @@
+//! Network systems implementation backed by the TCP network protocol.
+
+use crate::simulation::{
+    events::NetworkSimulationEvent,
+    requirements::DeliveryRequirement,
+    timing::{NetworkSimulationTime, NetworkSimulationTimeSystem},
+    transport::{
+        TransportResource, NETWORK_RECV_SYSTEM_NAME, NETWORK_SEND_SYSTEM_NAME,
+        NETWORK_SIM_TIME_SYSTEM_NAME,
+    },
+    Message,
+};
+use amethyst_core::{
+    bundle::SystemBundle,
+    ecs::{DispatcherBuilder, Read, System, World, Write},
+    shrev::EventChannel,
+};
+use amethyst_error::Error;
+use bytes::Bytes;
+use log::{error, warn};
+use std::{
+    collections::HashMap,
+    io::{self, Read as IORead, Write as IOWrite},
+    net::{SocketAddr, TcpListener, TcpStream},
+    ops::{Deref, DerefMut},
+};
+
+const CONNECTION_LISTENER_SYSTEM_NAME: &str = "connection_listener";
+
+/// Use this network bundle to add the various underlying tcp network systems to your game.
+pub struct TcpNetworkBundle {
+    listener: Option<TcpListener>,
+    recv_buffer_size_bytes: usize,
+}
+
+impl TcpNetworkBundle {
+    pub fn new(listener: Option<TcpListener>, recv_buffer_size_bytes: usize) -> Self {
+        Self {
+            listener,
+            recv_buffer_size_bytes,
+        }
+    }
+}
+
+impl<'a, 'b> SystemBundle<'a, 'b> for TcpNetworkBundle {
+    fn build(
+        self,
+        world: &mut World,
+        builder: &mut DispatcherBuilder<'_, '_>,
+    ) -> Result<(), Error> {
+        builder.add(
+            NetworkSimulationTimeSystem,
+            NETWORK_SIM_TIME_SYSTEM_NAME,
+            &[],
+        );
+        builder.add(
+            TcpConnectionListenerSystem,
+            CONNECTION_LISTENER_SYSTEM_NAME,
+            &[NETWORK_SIM_TIME_SYSTEM_NAME],
+        );
+        builder.add(
+            TcpNetworkSendSystem,
+            NETWORK_SEND_SYSTEM_NAME,
+            &[CONNECTION_LISTENER_SYSTEM_NAME],
+        );
+        builder.add(
+            TcpNetworkRecvSystem::new(),
+            NETWORK_RECV_SYSTEM_NAME,
+            &[CONNECTION_LISTENER_SYSTEM_NAME],
+        );
+        world.insert(TcpNetworkResource::new(
+            self.listener,
+            self.recv_buffer_size_bytes,
+        ));
+        Ok(())
+    }
+}
+
+/// System to listen for incoming connections and cache them to the resource.
+pub struct TcpConnectionListenerSystem;
```

Comment:
This can be private right?

Reply:
Not if we want to allow people to bind and set the tcplistener after initializing the system

Reply:
okay